### PR TITLE
Feature/reasoning layer

### PR DESCRIPTION
**Pull request by Dragos0000**

### CODE REVIEW

File: src/reasoning-layer-lib/elements-reasoning-layer.xsl [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        <xsl:variable name="attributeMultiplicityMax" select="$attribute/bounds/@upper"/>
+        <xsl:variable name="className" select="$attribute/../../@name"/>
+        <xsl:variable name="classURI" select="f:buildURIfromLexicalQName($className, fn:true())"/>
+        <xsl:variable name="datatypeURI"
+            select="f:buildURIfromLexicalQName('xsd:integer', fn:false())"/>
+        <xsl:variable name="attributeURI"
+            select="f:buildURIFromAttribute($attribute, fn:false(), fn:true())"/>
+        <owl:Class rdf:about="{$classURI}">
+            <rdfs:subClassOf>
+                <xsl:if test="$attributeMultiplicityMin != '*' and $attributeMultiplicityMax != '*'">
```
Comment (costezki) [inline src/reasoning-layer-lib/elements-reasoning-layer.xsl:121]: Do we need to conjunct these two conditions together?

File: src/reasoning-layer-lib/elements-reasoning-layer.xsl [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                            <owl:Restriction>
+                                <owl:onProperty rdf:resource="{$attributeURI}"/>
+                                <owl:maxCardinality rdf:datatype="{$datatypeURI}">
+                                    <xsl:value-of select="$attributeMultiplicityMax"/>
+                                </owl:maxCardinality>
+                            </owl:Restriction>
+                        </xsl:otherwise>
+                    </xsl:choose>
+                </xsl:if>
+                <xsl:if test="$attributeMultiplicityMin = '*'">
```
Comment (costezki) [inline src/reasoning-layer-lib/elements-reasoning-layer.xsl:147]: Will this logic fire independently or be conditioned by the above condition?

File: src/reasoning-layer-lib/elements-reasoning-layer.xsl [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                    <owl:Restriction>
+                        <owl:onProperty rdf:resource="{$attributeURI}"/>
+                        <owl:maxCardinality rdf:datatype="{$datatypeURI}">
+                            <xsl:value-of select="$attributeMultiplicityMax"/>
+                        </owl:maxCardinality>
+                    </owl:Restriction>
+                </xsl:if>
+                <xsl:if test="$attributeMultiplicityMax = '*'">
```
Comment (costezki) [inline src/reasoning-layer-lib/elements-reasoning-layer.xsl:155]: idem...

File: src/reasoning-layer-lib/elements-reasoning-layer.xsl [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+    <xd:doc>
+        <xd:desc>(Enumeration in reasnoning layer) . Concept instantiation in RDF/XML For an UML
+            enumeration, specify an equivalent class restriction covering the set of individuals
+            that are skos:inScheme of this enumeration.</xd:desc>
+    </xd:doc>
+    <xsl:template match="element[@xmi:type = 'uml:Enumeration']">
+        <xsl:variable name="enumerationURI" select="f:buildURIFromElement(., fn:true(), fn:true())"/>
+        <owl:Class rdf:about="{$enumerationURI}">
+            <rdfs:subClassOf rdf:resource="http://www.w3.org/2004/02/skos/core#Concept"/>
```
Comment (costezki) [inline src/reasoning-layer-lib/elements-reasoning-layer.xsl]: do not make it a skos:Concept for now. Comment out this line

File: src/reasoning-layer-lib/reasoning-layer-selectors.xsl [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        <xsl:apply-templates select="/xmi:XMI/xmi:Extension/elements"/>
+        <xsl:apply-templates select="/xmi:XMI/xmi:Extension/connectors"/>
+    </xsl:template>
+    
+    <xd:doc>
+        <xd:desc>
+            <xd:p> This template selects target element definitions</xd:p>
+        </xd:desc>
+    </xd:doc>
+    <xsl:template match="elements">
```
Comment (costezki) [inline src/reasoning-layer-lib/reasoning-layer-selectors.xsl:35]: sure we need to select only these two elements?

File: test/testData/ePO-CM-v2.0.1-2020-03-27_test.eap.xmi [hunk 6]
```diff
@@ -28198,7 +28198,7 @@
 						<initial/>
 						<documentation value="Any additional information about the recurrence of the Procurement.&#xA; &#xA;Additional Information:&#xA;For example estimated timing of the Procedure.&#xA;&#xA;&#xA;(WG approval 2019-01-16)"/>
 						<model ea_localid="850" ea_guid="{E8C2226A-C1ED-43ef-AE1E-2A685BD6A3BF}"/>
-						<properties type="Text" collection="false" static="0" duplicates="0" changeability="changeable"/>
```
Comment (costezki) [inline test/testData/ePO-CM-v2.0.1-2020-03-27_test.eap.xmi:28201]: do we handle both cases? of UML and XSD datatypes?

**Review by costezki — CHANGES_REQUESTED** [review]

File: src/reasoning-layer-lib/elements-reasoning-layer.xsl (hunk 4 above)
Comment (Dragos0000) [reply src/reasoning-layer-lib/elements-reasoning-layer.xsl]: did that

File: src/reasoning-layer-lib/reasoning-layer-selectors.xsl (hunk 5 above)
Comment (Dragos0000) [reply src/reasoning-layer-lib/reasoning-layer-selectors.xsl:35]: checked

File: src/reasoning-layer-lib/elements-reasoning-layer.xsl [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    </xd:doc>
+
+    <xsl:template name="attributeRange">
+        <xsl:param name="attribute"/>
+        <xsl:variable name="className" select="$attribute/../../@name"/>
+        <xsl:variable name="classURI" select="f:buildURIfromLexicalQName($className, fn:true())"/>
+        <xsl:variable name="attributeType" select="$attribute/properties/@type"/>
+        <xsl:variable name="attributeURI"
+            select="f:buildURIFromAttribute($attribute, fn:false(), fn:true())"/>
+        <xsl:if test="f:isAttributeTypeValidForDatatypeProperty($attribute)">
```
Comment (costezki) [inline src/reasoning-layer-lib/elements-reasoning-layer.xsl:75]: looks about right

**Pull request merged**